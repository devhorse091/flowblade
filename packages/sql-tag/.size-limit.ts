import type { SizeLimitConfig } from 'size-limit'

module.exports = [
  {
    name: 'Import * (ESM)',
    path: ['dist/index.mjs'],
    import: '*',
    limit: '1000B',
  },
  {
    name: 'Only { sql } (ESM)',
    path: ['dist/index.mjs'],
    import: '{ sql }',
    limit: '700B',
  },
] satisfies SizeLimitConfig;
