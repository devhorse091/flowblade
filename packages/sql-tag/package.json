{
  "name": "@flowblade/sql-tag",
  "description": "Raw sql template literal",
  "version": "0.0.1",
  "license": "MIT",
  "author": {
    "name": "Vanvelthem SÃ©bastien",
    "url": "https://github.com/belgattitude"
  },
  "homepage": "https://belgattitude.github.io/flowblade",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/belgattitude/flowblade.git",
    "directory": "packages/sql-tag"
  },
  "keywords": [],
  "sideEffects": false,
  "type": "module",
  "main": "./dist/index.cjs",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.cts",
  "exports": {
    ".": {
      "import": {
        "types": "./dist/index.d.ts",
        "default": "./dist/index.mjs"
      },
      "require": {
        "types": "./dist/index.d.cts",
        "default": "./dist/index.cjs"
      }
    },
    "./package.json": "./package.json"
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "?build-release": "When https://github.com/atlassian/changesets/issues/432 has a solution we can remove this trick",
    "build": "run clean && yarn run tsup",
    "build-release": "run build",
    "bench": "vitest bench --run",
    "bench-codspeed": "cross-env CODSPEED=1 vitest bench --run",
    "bench-watch": "vitest bench",
    "docgen": "run-s docgen-typedoc",
    "docgen-typedoc": "rimraf ./docs/api && typedoc --plugin typedoc-plugin-markdown --out ./docs/api",
    "check-dist": "run-s check-dist-esm check-dist-cjs",
    "check-dist-cjs": "es-check --not './dist/*.map.js' -v es2022 './dist/**/*.cjs'",
    "check-dist-esm": "es-check --not './dist/*.map.js' -v es2022 --module './dist/**/*.mjs'",
    "check-pub": "attw --pack && publint",
    "check-size": "size-limit",
    "clean": "rimraf --glob ./dist ./build ./coverage ./_release './tsconfig*.tsbuildinfo'",
    "dev": "tsup --watch",
    "fix-staged": "lint-staged --allow-empty",
    "lint": "eslint . --ext .ts,.tsx,.js,.jsx,.mjs,.cjs,.mts,.cts --cache --cache-location ../../.cache/eslint/core.eslintcache",
    "test": "vitest run",
    "test-unit": "vitest run",
    "test-e2e": "vitest --config=vitest.e2e.config.ts run",
    "test-unit-coverage": "vitest run --coverage",
    "test-unit-edge": "vitest run --environment edge-runtime",
    "test-unit-watch": "vitest --ui --api 4515",
    "typecheck": "tsc --project tsconfig.json --noEmit",
    "typecheck-build": "tsc --project tsconfig.build.json --noEmit",
    "ci-coverage-upload": "../../.github/scripts/download/codecov -F flowblade-core --dir ./coverage"
  },
  "devDependencies": {
    "@arethetypeswrong/cli": "0.17.2",
    "@belgattitude/eslint-config-bases": "6.17.0",
    "@codspeed/vitest-plugin": "4.0.0",
    "@edge-runtime/vm": "5.0.0",
    "@size-limit/file": "11.1.6",
    "@size-limit/webpack": "11.1.6",
    "@vitest/coverage-istanbul": "2.1.8",
    "@vitest/ui": "2.1.8",
    "browserslist": "4.24.3",
    "browserslist-to-esbuild": "2.1.1",
    "cross-env": "7.0.3",
    "es-check": "7.2.1",
    "es-toolkit": "1.30.1",
    "esbuild": "0.24.2",
    "eslint": "8.57.1",
    "execa": "9.5.2",
    "npm-run-all2": "7.0.2",
    "prettier": "3.4.2",
    "publint": "0.2.12",
    "rimraf": "6.0.1",
    "rollup": "4.29.1",
    "size-limit": "11.1.6",
    "sql-formatter": "15.4.8",
    "sql-template-tag": "5.2.1",
    "tsup": "8.3.5",
    "tsx": "4.19.2",
    "typedoc": "0.27.6",
    "typedoc-plugin-markdown": "4.3.3",
    "typescript": "5.7.2",
    "vite-tsconfig-paths": "5.1.4",
    "vitest": "2.1.8",
    "webpack": "5.97.1"
  },
  "engines": {
    "node": ">=18.17.0"
  }
}
