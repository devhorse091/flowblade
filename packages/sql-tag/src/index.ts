export { sql } from './sql';
export type { TaggedSql } from './types';
