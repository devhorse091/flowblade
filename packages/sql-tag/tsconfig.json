{
  "$schema": "https://json.schemastore.org/tsconfig",
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "target": "esnext",
    "module": "esnext",
    "moduleResolution": "bundler",
    "verbatimModuleSyntax": true,
    "strict": true,
    "noUncheckedIndexedAccess": true,
    "strictNullChecks": true,
    "exactOptionalPropertyTypes": false,
    "paths": {
      "@flowblade/sql-tag-format": ["../sql-tag-format/src/index"]
    },
    "types": ["vitest/globals"]
  },
  "exclude": ["**/node_modules", "**/.*/*", "dist", "coverage"],
  "include": [
    ".eslintrc.*",
    "**/*.ts",
    "**/*.tsx",
    "**/*.js",
    "**/*.jsx",
    "**/*.cjs",
    "**/*.mjs",
    "**/*.json"
  ]
}
