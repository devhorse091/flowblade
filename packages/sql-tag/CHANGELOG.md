# @flowblade/sql-tag

## 0.1.3

### Patch Changes

- [#262](https://github.com/belgattitude/flowblade/pull/262) [`493e1e8`](https://github.com/belgattitude/flowblade/commit/493e1e808b8435b7dbfa8ebc2a37d95d91710925) Thanks [@belgattitude](https://github.com/belgattitude)! - Rename TaggedSql type to SqlTag

## 0.1.2

### Patch Changes

- [#250](https://github.com/belgattitude/flowblade/pull/250) [`bbacdbf`](https://github.com/belgattitude/flowblade/commit/bbacdbff458c079df721db6241c3ff042c1c0e16) Thanks [@belgattitude](https://github.com/belgattitude)! - Fix last release

## 0.1.1

### Patch Changes

- [`546c69f`](https://github.com/belgattitude/flowblade/commit/546c69f7d52aa28ca0386b8076abc4ddd531afbb) - Republish

## 0.1.0

### Minor Changes

- [#237](https://github.com/belgattitude/flowblade/pull/237) [`d39d89c`](https://github.com/belgattitude/flowblade/commit/d39d89c88586fade87037081fa14d70e087b4017) Thanks [@belgattitude](https://github.com/belgattitude)! - Initial @flowblade/sql-tag release

  ## Install

  ```bash
  yarn add @flowblade/sql-tag
  ```

  ## Usage

  ```typescript
  import { sql } from "@flowblade/sql-tag";

  // ðŸ‘ˆ Unvalidated parameters
  const params = {
    country: "BE",
    users: ["John", "Doe"],
    ids: [1],
  };

  const query = sql<{
    id: number;
    username: string;
  }>`
     SELECT id, username FROM users
     WHERE country = ${params.country}           -- ðŸ‘ˆ simple
     AND username IN (${sql.join(params.users)}) -- ðŸ‘ˆ sql.join
  
     -- ðŸ‘‡ conditional clause with sql.empty
     ${params.ids.length > 0 ? sql`AND id IN (${sql.join(params.ids)})` : sql.empty}
  `;

  // query.sql === "SELECT id, username FROM users WHERE country = ? AND username IN (?, ?) AND id IN (?)";
  // query.values === ['BE', 'John', 'Doe', 1];
  ```
