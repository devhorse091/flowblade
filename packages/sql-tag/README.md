# @flowblade/sql-tag

Fast and lightweight ([~630B](#bundle-size)) sql template tag based on [sql-template-tag](https://github.com/blakeembrey/sql-template-tag).

[![npm](https://img.shields.io/npm/v/@flowblade/sql-tag?style=for-the-badge&label=Npm&labelColor=444&color=informational)](https://www.npmjs.com/package/@flowblade/sql-tag)
[![changelog](https://img.shields.io/static/v1?label=&message=changelog&logo=github&style=for-the-badge&labelColor=444&color=informational)](https://github.com/belgattitude/flowblade/blob/main/packages/sql-tag/CHANGELOG.md)
[![codecov](https://img.shields.io/codecov/c/github/belgattitude/flowblade?logo=codecov&label=Unit&flag=flowblade-sql-tag-unit&style=for-the-badge&labelColor=444)](https://app.codecov.io/gh/belgattitude/flowblade/tree/main/packages%2Fsql-tag)
[![bundles](https://img.shields.io/static/v1?label=&message=cjs|esm@treeshake&logo=webpack&style=for-the-badge&labelColor=444&color=informational)](https://github.com/belgattitude/flowblade/blob/main/packages/sql-tag/.size-limit.cjs)
[![node](https://img.shields.io/static/v1?label=Node&message=18%2b&logo=node.js&style=for-the-badge&labelColor=444&color=informational)](#compatibility)
[![browserslist](https://img.shields.io/static/v1?label=Browser&message=%3E96%25&logo=googlechrome&style=for-the-badge&labelColor=444&color=informational)](#compatibility)
[![size](https://img.shields.io/bundlephobia/minzip/@flowblade/sql-tag@latest?label=Max&style=for-the-badge&labelColor=444&color=informational)](https://bundlephobia.com/package/@flowblade/sql-tag@latest)
[![downloads](https://img.shields.io/npm/dm/@flowblade/sql-tag?style=for-the-badge&labelColor=444)](https://www.npmjs.com/package/@flowblade/sql-tag)
[![license](https://img.shields.io/npm/l/@flowblade/sql-tag?style=for-the-badge&labelColor=444)](https://github.com/belgattitude/flowblade/blob/main/LICENSE)

## Features

- üõ°Ô∏è&nbsp; Take advantage of [template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) to prevent sql injections.
- ü§≤&nbsp; Facilitate [query composition](#query-composition) and [conditional clauses](#conditionals).
- ü¶Ñ&nbsp; Separate actual sql from provided parameters.
- ‚ö°Ô∏è&nbsp; Minimal performance overhead.
- üìê&nbsp; Lightweight (less than [~700B](#bundle-size))
- ‚ôæÔ∏èÔ∏è&nbsp; Tested on [node 18-22, browser, workers and edge](#compatibility).
- üåç &nbsp; Available in ESM and CJS formats.

## Install

```bash
yarn add @flowblade/sql-tag
```
## Usage

### Basic

```typescript
import { sql } from '@flowblade/sql-tag';

// üëà User provided parameters
const params = {
    country: 'BE',
    users: ['John', 'Doe'],
};

const query = sql<{ // üëà optionally type the result
    id: number;
    username: string;
}>`
   SELECT id, username FROM users 
   WHERE country = ${params.country}           -- üëà simple param
   AND username IN (${sql.join(params.users)}) -- üëà array param
`;

// query.sql === "SELECT id, username FROM users WHERE country = ? AND username IN (?, ?)";
// query.values === ['BE', 'John', 'Doe'];
```

### Conditionals

```typescript
import { sql } from '@flowblade/sql-tag';

// üëà User provided parameters
const userIds = [1, 2];
const limit = 10;

const query = sql<{ // üëà optionally type the result
    id: number;
    username: string;
}>`
   SELECT id, username FROM users 
   WHERE 1=1 
   -- üëá alternative 2: with ternary operator and sql.empty
   ${userIds.length > 0 ? sql`AND id IN (${sql.join(userIds)})` : sql.empty}
   
   -- üëá alternative 2: with usage of sql.if helper
   ${sql.if(
     userIds.length,
     () => sql`AND id IN (${sql.join(userIds)})`
   )}    
   LIMIT ${limit}                 
`;

// query.sql === "SELECT id, username FROM users WHERE 1=1 AND id IN (?, ?) LIMIT ?";
// query.values === [1, 2, 10];
```


### Query composition

You can nest any query into another one.

```typescript
import {sql} from '@flowblade/sql-tag';

const getSqlUserCountByCountries = (minUsers: number) => sql`
  SELECT 
    c.name as country_name, 
    count(u.id) as user_count 
  FROM country AS c INNER JOIN user u 
  ON c.id = u.country_id
  GROUP BY c.name
  HAVING count(u.id) > ${minUsers}
`;

const compression: 'zstd' | 'snappy' | 'gzip' = 'zstd';

// Example base on DuckDb COPY statement
// but you can nest into CTE, table aliases, subqueries etc...
const query = sql`
    COPY
    (${getSqlUserCountByCountries(23)})
    TO 'usercount_by_countries.parquet'
    (FORMAT 'parquet', COMPRESSION ${compression}, ROW_GROUP_SIZE 100000);
`;

console.log(query.values); // [23, 'zstd']
console.log(query.sql);    // "COPY (SELECT...."
```

### Bulk inserts

Ease bulk inserts/merge from multi rows arrays.

```typescript
import { sql } from '@flowblade/sql-tag';

const insert = sql`
   INSERT INTO product (name, price, stock, status) 
   VALUES ${sql.bulk([
     ['Laptop', 999.99, 50, 'active'],
     ['Keyboard', 79.99, 100, 'active'],
   ])}
  `;

const { text, sql, statement, values } = insert;

insert.text;   //=> "INSERT INTO product (name, price, stock, status) VALUES ($1,$2,$3,$4),($5,$6,$7,$8)"
insert.sql;    //=> "INSERT INTO product (name, price, stock, status) VALUES (?,?,?,?),(?,?,?,?),(?,?,?,?)"
insert.values; //=> ["Laptop", 999.99, 50, "active", "Keyboard", 79.99, 100, "active"]
```

// Example with pglite
const result = await db.query(text, values, {});
```

## Methods

| Helpers       | Description                               | Example                                              |
|---------------|-------------------------------------------|------------------------------------------------------|
| sql.unsafeRaw | Allow to pass unsafe values in the query. | `ORDER BY ${sql.unsafeRaw('name desc')}`             |
| sql.empty     | Helper to represent empty string.         | `${isTrue ? sql'1=1' : sql.empty}`                   |
| sql.join      | Join array values with optional separator | `AND id IN ${sql.join(['1', '3'])`                   |
| sql.if        | Conditionally add a statement             | `AND ${sql.if(true, () => sql'deleted_at is null')}` |
| sql.bulk      | Ease bulk inserts                         |                                                      |

## Recipes

### Advanced examplew with transact-sql (mssql)

```typescript
const products = Array.from({ length: 1000 }, (_, i) => ({
  id: i,
  productName: `Product ${i}`,
}));

const sqlRaw = sql<GetAdvancedResult>`
  -- TRANSACT-SQL
  DECLARE @Products NVARCHAR(MAX); -- WARNING LIMIT TO 2GB
  SET @Products = ${JSON.stringify(products)};

  -- DDL (# prefix is equivalent to CREATE TEMPORATY TABLE on other db)
  CREATE TABLE #products (
    productId INT,
    name NVARCHAR(255),
  );
  
  -- INSERT
  INSERT INTO #products (productId, productName)
     SELECT productId, productName
       FROM OPENJSON(@InitialData) WITH (
           id INT,           
           name NVARCHAR(255)
       );
          
  -- SELECT
  SELECT TOP ${sql.unsafeRaw(String(limit))} id, name
    FROM #products
    WHERE id > ${sql.unsafeRaw(String(offset))}
    ORDER BY id; 
`;

```

## Credits

This package won't be possible without the great work of [Blake Embrey sql-template-tag](https://github.com/blakeembrey/sql-template-tag).

Some notable differences:

- [x] Named export for sql: `import {sql} from '@flowblade/sql-tag'`.
- [x] Possibility to type the result of the query (ie `sql<Row>`).
- [x] Utility functions (join...) are directly available from the sql tag.
- [x] Add `sql.if` helper to conditionally add a statement.
- [x] Rename `sqlRaw` to `sql.unsafeRaw` to prevent misuse.

## Bundle size

Bundle size is tracked by a [size-limit configuration](https://github.com/belgattitude/flowblade/blob/main/packages/sql-tag/.size-limit.ts)

| Scenario (esm)                                              | Size (compressed) |
|-------------------------------------------------------------|------------------:|
| `import { sql } from '@flowblade/sql-tag`                   |            ~ 630B |

## Compatibility

| Level      | CI | Description                                                                                                                                                                                                                                                                                                                                    |
|------------|----|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|  
| Node       | ‚úÖ  | CI for 18.x, 20.x & 22.x.                                                                                                                                                                                                                                                                                                                      |
| Browser    | ‚úÖ  | Tested with latest chrome (vitest/playwright)                                                                                                                                                                                                                                                                                            |
| Browsers   | ‚úÖ  | [> 96%](https://browserslist.dev/?q=ZGVmYXVsdHMsIGNocm9tZSA%2BPSA5NixmaXJlZm94ID49IDkwLGVkZ2UgPj0gMTksc2FmYXJpID49IDEyLGlvcyA%2BPSAxMixvcGVyYSA%2BPSA3Nw%3D%3D) on 07/2024. Mins to [Chrome 96+, Firefox 90+, Edge 19+, iOS 12+, Safari 12+, Opera 77+](https://github.com/belgattitude/flowblade/blob/main/packages/sql-tag/.browserslistrc) |
| Edge       | ‚úÖ  | Ensured on CI with [@vercel/edge-runtime](https://github.com/vercel/edge-runtime).                                                                                                                                                                                                                                                       | 
| Cloudflare | ‚úÖ  | Ensured with @cloudflare/vitest-pool-workers (see [wrangler.toml](https://github.com/belgattitude/flowblade/blob/main/devtools/vitest/wrangler.toml)                                                                                                                                                                                         |
| Typescript | ‚úÖ  | TS 5.0 + / [are-the-type-wrong](https://github.com/arethetypeswrong/arethetypeswrong.github.io) checks on CI.                                                                                                                                                                                                                                  |
| ES2022     | ‚úÖ  | Dist files checked with [es-check](https://github.com/yowainwright/es-check)                                                                                                                                                                                                                                                                   |
| Performance| ‚úÖ  | Monitored with [codspeed.io](https://codspeed.io/belgattitude/flowblade)                                                                                                                                                                                                                                                                      |

## Contributors

Contributions are welcome. Have a look to the [CONTRIBUTING](https://github.com/belgattitude/flowblade/blob/main/CONTRIBUTING.md) document.

## Sponsors

[Sponsor](<[sponsorship](https://github.com/sponsors/belgattitude)>), [coffee](<(https://ko-fi.com/belgattitude)>),
or star ‚Äì All is spent for quality time with loved ones. Thanks ! üôè‚ù§Ô∏è

### Special thanks to

<table>
  <tr>
    <td>
      <a href="https://www.jetbrains.com/?ref=belgattitude" target="_blank">
         <img width="65" src="https://asset.brandfetch.io/idarKiKkI-/id53SttZhi.jpeg" alt="Jetbrains logo" />
      </a>
    </td>
    <td>
      <a href="https://www.embie.be/?ref=belgattitude" target="_blank">
        <img width="65" src="https://avatars.githubusercontent.com/u/98402122?s=200&v=4" alt="Jetbrains logo" />    
      </a>
    </td>
  </tr>
  <tr>
    <td align="center">
      <a href="https://www.jetbrains.com/?ref=belgattitude" target="_blank">JetBrains</a>
    </td>
    <td align="center">
      <a href="https://www.embie.be/?ref=belgattitude" target="_blank">Embie.be</a>
    </td>
   </tr>
</table>

## License

MIT ¬© [S√©bastien Vanvelthem](https://github.com/belgattitude) and contributors.
