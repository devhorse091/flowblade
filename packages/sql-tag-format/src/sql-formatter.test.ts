import { sql } from '@flowblade/sql-tag';

import {
  SqlFormatter,
  type SqlFormatterDialect,
  type SqlFormatterOptions,
} from './sql-formatter';

describe('Sql tag formatter', () => {
  const dialects = [
    'bigquery',
    'db2',
    'db2i',
    'hive',
    'mariadb',
    'mysql',
    'n1ql',
    'plsql',
    'postgresql',
    'redshift',
    'singlestoredb',
    'snowflake',
    'spark',
    'sql',
    'sqlite',
    'tidb',
    'transactsql',
    'trino',
    'tsql',
  ] as const satisfies SqlFormatterDialect[];

  const formatterOptions = {
    keywordCase: 'preserve',
    identifierCase: 'preserve',
    dataTypeCase: 'preserve',
    functionCase: 'preserve',
    logicalOperatorNewline: 'before',
    expressionWidth: 50,
    linesBetweenQueries: 1,
    denseOperators: false,
    newlineBeforeSemicolon: false,
    useTabs: false,
    tabWidth: 2,
  } as const satisfies SqlFormatterOptions;

  describe('formatOrNull', () => {
    it.each(dialects)('should format the query for %s', (dialect) => {
      const sqlFormatter = new SqlFormatter(dialect, formatterOptions);
      const formatted = sqlFormatter.formatOrThrow(
        sql`SELECT * FROM table WHERE id = 1`
      );
      expect(formatted).toMatchSnapshot();
    });
    describe('formatOrThrow', () => {
      it.each(dialects)('should format the query for %s', (dialect) => {
        const sqlFormatter = new SqlFormatter(dialect, formatterOptions);
        const formatted = sqlFormatter.formatOrThrow(
          sql`SELECT * FROM table WHERE id = 1`
        );
        expect(formatted).toMatchSnapshot();
      });
    });
  });
  describe('Invalid query', () => {
    const invalidSql = sql`SELECT [A] FROM TABLE WHERE 1 = TABLE`;
    describe('formatOrNull', () => {
      it('should return null when the query is invalid', () => {
        const sqlFormatter = new SqlFormatter('sql', formatterOptions);
        const formatted = sqlFormatter.formatOrNull(invalidSql);
        expect(formatted).toBeNull();
      });
    });
    describe('formatOrThrow', () => {
      it('should return null when the query is invalid', () => {
        const sqlFormatter = new SqlFormatter('sql', formatterOptions);

        expect(() => {
          sqlFormatter.formatOrThrow(invalidSql);
        }).toThrow();
      });
    });
  });
  describe('Complex transact sql query', () => {
    it('should format the query', () => {
      const sqlFormatter = new SqlFormatter('transactsql', formatterOptions);
      const sqlRaw = sql`
          -- TRANSACT-SQL
          DECLARE @InitialData NVARCHAR(MAX); 
          DECLARE @ProductToUpdate NVARCHAR(MAX);        
          SET @InitialData = ${JSON.stringify([])};
          SET @ProductToUpdate = ${JSON.stringify([])};
      
          -- DDL
          CREATE TABLE #correctedProducts (
            productId NVARCHAR(255),
            countryId NVARCHAR(10),
            productName NVARCHAR(255),
            createdAt DATETIME DEFAULT GETDATE(),
            updatedAt DATETIME DEFAULT GETDATE(),
          );
          
          -- INSERT
          INSERT INTO #correctedProducts (productId, countryId, productName)
             SELECT productId, countryId, productName
               FROM OPENJSON(@InitialData) WITH (
                   productId NVARCHAR(255), 
                   countryId NVARCHAR(255),
                   productName NVARCHAR(255)
               );
                
          -- FROM HERE I AM IN A SITUATION WHERE THE TABLE Is FILLED
          UPDATE T
          SET productName = tNewData.productName
          FROM (SELECT productId, countryId, productName 
                FROM OPENJSON(@ProductToUpdate) 
                WITH (
                    productId NVARCHAR(255), 
                    countryId NVARCHAR(255),
                    productName NVARCHAR(255))
                ) AS tNewData
          INNER JOIN #correctedProducts AS T
          ON tNewData.productId = T.productId and T.countryId = tNewData.countryId;    
          
          -- SELECT
          SELECT TOP ${sql.unsafeRaw(String(10))} productId, countryId, productName, createdAt, updatedAt 
          FROM #correctedProducts;
      `;

      const formatted = sqlFormatter.formatOrThrow(sqlRaw, {
        options: {
          /**
           * @see https://github.com/sql-formatter-org/sql-formatter/blob/master/docs/paramTypes.md#option-value
           */
          paramTypes: {
            positional: true,
          },
        },
      });

      expect(formatted).toMatchSnapshot();
    });
  });
});
