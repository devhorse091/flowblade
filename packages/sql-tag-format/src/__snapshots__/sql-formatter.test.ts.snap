// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sql tag formatter > Complex transact sql query > should format the query 1`] = `
"-- TRANSACT-SQL
DECLARE @InitialData NVARCHAR(MAX);

DECLARE @ProductToUpdate NVARCHAR(MAX);

SET
  @InitialData = ?;

SET
  @ProductToUpdate = ?;

-- DDL
CREATE TABLE #correctedProducts (
  productId NVARCHAR(255),
  countryId NVARCHAR(10),
  productName NVARCHAR(255),
  createdAt DATETIME DEFAULT GETDATE(),
  updatedAt DATETIME DEFAULT GETDATE(),
);

-- INSERT
INSERT INTO
  #correctedProducts (productId, countryId, productName)
SELECT
  productId,
  countryId,
  productName
FROM
  OPENJSON (@InitialData)
WITH
  (
    productId NVARCHAR(255),
    countryId NVARCHAR(255),
    productName NVARCHAR(255)
  );

-- FROM HERE I AM IN A SITUATION WHERE THE TABLE Is FILLED
UPDATE T
SET
  productName = tNewData.productName
FROM
  (
    SELECT
      productId,
      countryId,
      productName
    FROM
      OPENJSON (@ProductToUpdate)
    WITH
      (
        productId NVARCHAR(255),
        countryId NVARCHAR(255),
        productName NVARCHAR(255)
      )
  ) AS tNewData
  INNER JOIN #correctedProducts AS T ON tNewData.productId = T.productId
  and T.countryId = tNewData.countryId;

-- SELECT
SELECT
  TOP 10 productId,
  countryId,
  productName,
  createdAt,
  updatedAt
FROM
  #correctedProducts;"
`;

exports[`Sql tag formatter > formatOrNull > formatOrThrow > should format the query for bigquery 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > formatOrThrow > should format the query for db2 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > formatOrThrow > should format the query for db2i 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > formatOrThrow > should format the query for hive 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > formatOrThrow > should format the query for mariadb 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > formatOrThrow > should format the query for mysql 1`] = `
"SELECT
  *
FROM
table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > formatOrThrow > should format the query for n1ql 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > formatOrThrow > should format the query for plsql 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > formatOrThrow > should format the query for postgresql 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > formatOrThrow > should format the query for redshift 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > formatOrThrow > should format the query for singlestoredb 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > formatOrThrow > should format the query for snowflake 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > formatOrThrow > should format the query for spark 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > formatOrThrow > should format the query for sql 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > formatOrThrow > should format the query for sqlite 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > formatOrThrow > should format the query for tidb 1`] = `
"SELECT
  *
FROM
table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > formatOrThrow > should format the query for transactsql 1`] = `
"SELECT
  *
FROM
table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > formatOrThrow > should format the query for trino 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > formatOrThrow > should format the query for tsql 1`] = `
"SELECT
  *
FROM
table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > should format the query for bigquery 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > should format the query for db2 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > should format the query for db2i 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > should format the query for hive 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > should format the query for mariadb 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > should format the query for mysql 1`] = `
"SELECT
  *
FROM
table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > should format the query for n1ql 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > should format the query for plsql 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > should format the query for postgresql 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > should format the query for redshift 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > should format the query for singlestoredb 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > should format the query for snowflake 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > should format the query for spark 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > should format the query for sql 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > should format the query for sqlite 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > should format the query for tidb 1`] = `
"SELECT
  *
FROM
table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > should format the query for transactsql 1`] = `
"SELECT
  *
FROM
table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > should format the query for trino 1`] = `
"SELECT
  *
FROM
  table
WHERE
  id = 1"
`;

exports[`Sql tag formatter > formatOrNull > should format the query for tsql 1`] = `
"SELECT
  *
FROM
table
WHERE
  id = 1"
`;
